tests/e2e/loadbalancer.go

```diff
@@ -77,225 +72,276 @@ var _ = Describe("[cloud-provider-aws-e2e] loadbalancer", func() {
-			Annotations: map[string]string{
-				annotationLBType: "nlb",
-				"service.beta.kubernetes.io/aws-load-balancer-internal": "true",
+			name:           "NLB internal should be reachable with hairpinning traffic",
+			resourceSuffix: "hp-nlb-int",
+			extraAnnotations: map[string]string{
+				annotationLBType:     "nlb",
+				annotationLBInternal: "true",
 			},
-			PostConfigService: func(cfg *configServiceLB, svc *v1.Service, nodeDiscovery ClusterNodeDiscovery) {
-				if svc.Annotations == nil {
-					svc.Annotations = map[string]string{}
+			listenerCount: 1,
+			hookPostServiceConfig: func(cfg *e2eTestConfig) {
+				framework.Logf("running hook post-service-config patching service annotations to enforce LB pins/selects target to a single node: kubernetes.io/hostname=%s", cfg.nodeSingleSample)
+				if cfg.svc.Annotations == nil {
+					cfg.svc.Annotations = map[string]string{}
 				}
-				svc.Annotations[annotationLBTargetNodeLabels] = fmt.Sprintf("kubernetes.io/hostname=%s", nodeDiscovery.SingleWorker)
-				framework.Logf("Using service annotations: %v", svc.Annotations)
+				cfg.svc.Annotations[annotationLBTargetNodeLabels] = fmt.Sprintf("kubernetes.io/hostname=%s", cfg.nodeSingleSample)
 			},
-			HookInClusterTestReachableHTTP: true,
-			RequireAffinity:                true,
-			SkipTestFailure:                true,
+			overrideTestRunInClusterReachableHTTP: true,
+			requireAffinity:                       true,
+			skipTestFailure:                       true,
 		},
 	}
 
 	serviceNameBase := "lbconfig-test"
 	for _, tc := range cases {
-		It(tc.Name, func() {
-			nodeDiscovery := discoverClusterWorkerNode(cs)
+		It(tc.name, func() {
+			By("setting up test environment and discovering worker nodes")
+			e2e := newE2eTestConfig(cs)
+			e2e.discoverClusterWorkerNode()
+			framework.Logf("[SETUP] Test case: %s", tc.name)
+			framework.Logf("[SETUP] Worker nodes discovered: %d nodes, selector: %s, sample node: %s", e2e.nodeCount, e2e.nodeSelector, e2e.nodeSingleSample)
+
 			loadBalancerCreateTimeout := e2eservice.GetServiceLoadBalancerCreationTimeout(cs)
-			framework.Logf("Running tests against AWS with timeout %s", loadBalancerCreateTimeout)
+			framework.Logf("[CONFIG] AWS load balancer timeout: %s", loadBalancerCreateTimeout)
 
-			// Create Configuration
+			By("building service configuration with annotations")
 			serviceName := serviceNameBase
-			if len(tc.ResourceSuffix) > 0 {
-				serviceName = serviceName + "-" + tc.ResourceSuffix
+			if len(tc.resourceSuffix) > 0 {
+				serviceName = serviceName + "-" + tc.resourceSuffix
 			}
-			framework.Logf("namespace for load balancer conig test: %s", ns.Name)
-
-			By("creating a TCP service " + serviceName + " with type=LoadBalancerType in namespace " + ns.Name)
-			lbConfig := newConfigServiceLB()
-			lbConfig.LBJig = e2eservice.NewTestJig(cs, ns.Name, serviceName)
+			framework.Logf("[CONFIG] Service name: %s, namespace: %s", serviceName, ns.Name)
+			e2e.newConfigServiceLB()
```

Comment:
i think changing this function to not have a return value makes the readability a little worse, since the function is now hiding a side-effect. why change this to not return the struct?

Reply:
good question. The main idea was to move from self-contained service configuration for the test case (type `configServiceLB`) to a wider test case configuration container(type `e2eTestConfig`) which holds those service configurations, as well more attributes used to the test case variability (such as node selector, count and single-hostname, etc).

imho we can move that "setter" function to the `newE2eTestConfig()` as it is just setting initial values. Is that makes sense to you?

Reply:
i think putting it in the initializer makes good sense and follows what i might expect.